src/wallet/wallet2.cpp

```diff
@@ -4315,31 +4542,80 @@ std::vector<wallet2::pending_tx> wallet2::create_transactions_2(std::vector<cryp
     THROW_WALLET_EXCEPTION_IF(0 == dt.amount, error::zero_destination);
     needed_money += dt.amount;
     LOG_PRINT_L2("transfer: adding " << print_money(dt.amount) << ", for a total of " << print_money (needed_money));
-    THROW_WALLET_EXCEPTION_IF(needed_money < dt.amount, error::tx_sum_overflow, dsts, 0, m_testnet);
+    THROW_WALLET_EXCEPTION_IF(needed_money < dt.amount, error::tx_sum_overflow, dsts, 0, m_testnet, is_subaddress);
   }
 
   // throw if attempting a transaction with no money
   THROW_WALLET_EXCEPTION_IF(needed_money == 0, error::zero_destination);
 
-  // gather all our dust and non dust outputs
+  std::map<uint32_t, uint64_t> balance_per_subaddr = unlocked_balance_per_subaddress();
+
+  if (subaddr_indices.empty()) // "index=<N1>[,<N2>,...]" wasn't specified -> use all the indices with non-zero unlocked bakance
+  {
+    for (auto& i : balance_per_subaddr)
+      subaddr_indices.insert(i.first);
+  }
+  const std::set<uint32_t> subaddr_indices_in = subaddr_indices;
+  subaddr_indices.clear();
+
+  // early out if we know we can't make it anyway
+  // we could also check for being within FEE_PER_KB, but if the fee calculation
+  // ever changes, this might be missed, so let this go through
+  uint64_t balance_subtotal = 0;
+  for (uint32_t index_minor : subaddr_indices_in)
+    balance_subtotal += balance_per_subaddr[index_minor];
+  THROW_WALLET_EXCEPTION_IF(needed_money > balance_subtotal, error::not_enough_money,
+    balance_subtotal, needed_money, 0);
+
+  // if one or more subaddresses have sufficient balance, choose one randomly (with index=0 being chosen last)
+  for (uint32_t index_minor : subaddr_indices_in)
+  {
+    if (balance_per_subaddr[index_minor] >= needed_money)
+      subaddr_indices.insert(index_minor);
+  }
+  if (subaddr_indices.size() > 0)
+  {
+    if (subaddr_indices.count(0) == 1 && subaddr_indices.size() > 1)
+      subaddr_indices.erase(0);
+    auto i = subaddr_indices.begin();
+    std::advance(i, crypto::rand<size_t>() % subaddr_indices.size());
+    subaddr_indices = { *i };
+  }
+  // otherwise, sort indices in the order of descending balances, and pick top n indices with subtotal balance sufficient for the needed money
```

Comment:
One could argue there are ways to improve upon this. E.g. when sending to a particular recipient a second time, to attempt to re-use funds from the same subaddress used last time and from the change returned last time.

This starts to get complicated though and the user can't rely on this to protect their privacy. 

Therefore I'm happy with @kenshi84 's implementation here, and in the long term perhaps we can revisit this when taking into consideration Surae/Sarang's recommendations.